### ISSUE TITLE
Support apiKey[FEATURE]

### ISSUE THREAD

**Issue by fpolowood** [epam-cross-platform-lab/swagger-dart-code-generator#290]
My API requires an apiKey to be sent with almost every call, swagger supports it via the:

```dart
"securityDefinitions": {
  "api_key": {
    "type": "apiKey",
    "name": "api_key",
    "in": "header"
  }
}
```
and all the calls have it as:

```dart
"/companyInformation": {
  "get": {
    ...
    "security": [{"api_key":[]}]
  }
}
```

When generating the code with the swagger tools, it allows me to pass the parameters for auth to the api_client instance and then they get injected into any of the calls that require it (in dart and js, haven't checked others).

An easy fix for me was to pass the api_key on the creation of the Swagger.create() call and then inject it via the HeadersInterceptor to every call. 
```dart
final newClient = ChopperClient(
  ...
  interceptors: [
    HeadersInterceptor({'api_key': theKey}),
  ],
  baseUrl: 'http://localhost:3001/v1');
```
The downside of this approach is that, if the API changes I will need to update the files manually every time.  Also, the headers are being applied indiscriminately to all calls, which in my case is fine, but it might not be the correct solution for everyone.

The generic fix should use the swagger file to determine the name of the apiKe, which calls require it and build a custom interceptor to inject the headers.

**Comment by Vovanella95** [comment]
Hi @fpolowood 
Sounds like cool feature! I will try to implement it soon!

**Comment by Vovanella95** [comment]
@fpolowood  

Could you please provide real swagger file with such definition and expected generated file? What if some request doesn't have  "security": [{"api_key":[]}]?

**Comment by fpolowood** [comment]
```json
{
  "swagger": "2.0",
  "info": {
    "description": "MyStore API",
    "version": "1.0.0",
    "title": "MyStore",
    "termsOfService": "",
    "contact": {
      "email": "felipe@migrandama.com"
    },
    "license": {
      "name": "MIT"
    }
  },
  "host": "localhost:3001",
  "basePath": "/v1",
  "schemes": [
    "http"
  ],
  "paths": {
    "/task": {
      "post": {
        "tags": [
          "task"
        ],
        "summary": "Add a new task",
        "description": "",
        "operationId": "addTask",
        "consumes": [
          "application/json"
        ],
        "produces": [
          "application/json"
        ],
        "parameters": [
          {
            "in": "body",
            "name": "body",
            "description": "Task object that needs to be added to the store",
            "required": true,
            "schema": {
              "$ref": "#/definitions/Task"
            }
          }
        ],
        "responses": {
          "405": {
            "description": "Invalid input"
          }
        },
        "security": [
          {
            "api_key": []
          }
        ]
      },
      "put": {
        "tags": [
          "task"
        ],
        "summary": "Update an existing task",
        "description": "",
        "operationId": "updateTask",
        "consumes": [
          "application/json"
        ],
        "produces": [
          "application/json"
        ],
        "parameters": [
          {
            "in": "body",
            "name": "body",
            "description": "Task object that needs to be added to the store",
            "required": true,
            "schema": {
              "$ref": "#/definitions/Task"
            }
          }
        ],
        "responses": {
          "400": {
            "description": "Invalid ID supplied"
          },
          "404": {
            "description": "Task not found"
          },
          "405": {
            "description": "Validation exception"
          }
        },
        "security": [
          {
            "api_key": []
          }
        ]
      }
    },
    "/task/findByStatus": {
      "get": {
        "tags": [
          "task"
        ],
        "summary": "Finds Tasks by status",
        "description": "Multiple status values can be provided with comma separated strings",
        "operationId": "findTasksByStatus",
        "produces": [
          "application/json"
        ],
        "parameters": [
          {
            "name": "status",
            "in": "query",
            "description": "Status values that need to be considered for filter",
            "required": true,
            "type": "array",
            "items": {
              "type": "string",
              "default": "new",
              "enum": [
                "new",
                "active",
                "done"
              ]
            },
            "collectionFormat": "multi"
          }
        ],
        "responses": {
          "200": {
            "description": "successful operation",
            "schema": {
              "type": "array",
              "items": {
                "$ref": "#/definitions/Task"
              }
            }
          },
          "400": {
            "description": "Invalid status value"
          }
        },
        "security": [
          {
            "api_key": []
          }
        ]
      }
    },
    "/task/{taskId}": {
      "get": {
        "tags": [
          "task"
        ],
        "summary": "Find task by Id",
        "description": "Returns a single task",
        "operationId": "getTaskById",
        "produces": [
          "application/json"
        ],
        "parameters": [
          {
            "name": "taskId",
            "in": "path",
            "description": "Id of task to return",
            "required": true,
            "type": "integer",
            "format": "int64"
          }
        ],
        "responses": {
          "200": {
            "description": "successful operation",
            "schema": {
              "$ref": "#/definitions/Task"
            }
          },
          "400": {
            "description": "Invalid Id supplied"
          },
          "404": {
            "description": "Task not found"
          }
        },
        "security": [
          {
            "api_key": []
          }
        ]
      },
      "post": {
        "tags": [
          "task"
        ],
        "summary": "Updates a task in the store with form data",
        "description": "",
        "operationId": "updateTaskWithForm",
        "consumes": [
          "application/x-www-form-urlencoded"
        ],
        "produces": [
          "application/json"
        ],
        "parameters": [
          {
            "name": "taskId",
            "in": "path",
            "description": "Id of task that needs to be updated",
            "required": true,
            "type": "integer",
            "format": "int64"
          },
          {
            "name": "name",
            "in": "formData",
            "description": "Updated name of the task",
            "required": false,
            "type": "string"
          },
          {
            "name": "status",
            "in": "formData",
            "description": "Updated status of the task",
            "required": false,
            "type": "string"
          }
        ],
        "responses": {
          "405": {
            "description": "Invalid input"
          }
        },
        "security": [
          {
            "api_key": []
          }
        ]
      },
      "delete": {
        "tags": [
          "task"
        ],
        "summary": "Deletes a task",
        "description": "",
        "operationId": "deleteTask",
        "produces": [
          "application/json"
        ],
        "parameters": [
          {
            "name": "taskId",
            "in": "path",
            "description": "Task id to delete",
            "required": true,
            "type": "integer",
            "format": "int64"
          }
        ],
        "responses": {
          "400": {
            "description": "Invalid id supplied"
          },
          "404": {
            "description": "Task not found"
          }
        },
        "security": [
          {
            "api_key": []
          }
        ]
      }
    },
    "/task/{taskId}/activate": {
      "put": {
        "tags": [
          "task"
        ],
        "summary": "Activates the task",
        "description": "Changes status to active and timestamps the DateActivated",
        "operationId": "activateTask",
        "produces": [
          "application/json"
        ],
        "parameters": [
          {
            "name": "taskId",
            "in": "path",
            "description": "Task id to activate",
            "required": true,
            "type": "integer",
            "format": "int64"
          }
        ],
        "responses": {
          "200": {
            "description": "Task activated"
          },
          "400": {
            "description": "Invalid taskId"
          }
        },
        "security": [
          {
            "api_key": []
          }
        ]
      }
    },
    "/task/{taskId}/complete": {
      "put": {
        "tags": [
          "task"
        ],
        "summary": "Completes the task",
        "description": "Changes status to done and timestamps the DateCompleted",
        "operationId": "completeTask",
        "produces": [
          "application/json"
        ],
        "parameters": [
          {
            "name": "taskId",
            "in": "path",
            "description": "Task id to complete",
            "required": true,
            "type": "integer",
            "format": "int64"
          }
        ],
        "responses": {
          "200": {
            "description": "Task completed"
          },
          "400": {
            "description": "Invalid taskId"
          }
        },
        "security": [
          {
            "api_key": []
          }
        ]
      }
    },
    "/companyInformation": {
      "get": {
        "tags": [
          "company information"
        ],
        "summary": "Get the company information",
        "operationId": "getCompanyInformation",
        "produces": [
          "application/json"
        ],
        "responses": {
          "200": {
            "description": "Success"
          }
        }
      }
    }
  },
  "securityDefinitions": {
    "api_key": {
      "type": "apiKey",
      "name": "api_key",
      "in": "header"
    }
  },
  "definitions": {
    "TaskCategory": {
      "type": "object",
      "properties": {
        "taskCategoryId": {
          "type": "integer",
          "format": "int64"
        },
        "taskCategoryName": {
          "type": "string"
        }
      }
    },
    "Task": {
      "type": "object",
      "required": [
        "TaskName"
      ],
      "properties": {
        "TaskID": {
          "type": "integer",
          "format": "int64"
        },
        "TaskName": {
          "type": "string"
        },
        "TaskCategory": {
          "$ref": "#/definitions/TaskCategory"
        },
        "Details": {
          "type": "string"
        },
        "Status": {
          "type": "string",
          "description": "Task status in the store",
          "enum": [
            "new",
            "active",
            "done"
          ]
        }
      }
    }
  }
}
```

**Comment by Vovanella95** [comment]
@fpolowood  what should be generated for requests without [{"api_key":[]}]? If we add an interceptor - this header will be applied to all requests.

**Comment by Vovanella95** [comment]
@fpolowood  OK let me implement it and see, what happens

**Comment by fpolowood** [comment]
Even if you are unable to implement all the use cases.. this diagram should give you an idea of the flow
![image](https://user-images.githubusercontent.com/9532726/147094889-c6a2b59f-a70a-46de-8d2d-a149d93d2b8e.png)

**Comment by Vovanella95** [comment]
Hi @fpolowood I've implemented this in 2.3.2 version. Please check and let us know is everything OK
